#### JEWELRY PRICE OPTIMIZATION PROJECT
This Machine Learning project will be carried out using the CRISP-DM Methodolody

##### Data understanding and exploration

In [2]:
# import libraries to work with
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# set visualization theme
sns.set()

import warnings
warnings.simplefilter(action='ignore', category=pd.errors.SettingWithCopyWarning)

In [5]:
# load the dataset
df = pd.read_csv(r"C:\Users\User\Desktop\jp project\data\Jewelry_Dataset.csv")
df

2018-12-01 11:40:29 UTC  1924719191579951782  1842195256808833386  1  \
0      2018-12-01 17:38:31 UTC  1924899396621697920  1806829193678291446  1   
1      2018-12-02 13:53:42 UTC  1925511016616034733  1842214461889315556  1   
2      2018-12-02 17:44:02 UTC  1925626951238681511  1835566849434059453  1   
3      2018-12-02 21:30:19 UTC  1925740842841014667  1873936840742928865  1   
4      2018-12-02 22:09:34 UTC  1925760595336888995  1835566854827934449  1   
...                        ...                  ...                  ... ..   
95905  2021-12-01 09:47:37 UTC  2719016587653808559  1515966223597820877  1   
95906  2021-12-01 09:50:37 UTC  2719018098886705639  1956663830729720074  1   
95907  2021-12-01 09:52:06 UTC  2719018846999544322  1956663848219968336  1   
95908  2021-12-01 09:55:35 UTC  2719020596963836492  1835566854106513435  1   
95909  2021-12-01 09:59:07 UTC  2719022379232658075  1956663846156370429  1   

       1806829201890738522   jewelry.earring             0  561.51  \
0             1.806829e+18               NaN           NaN  212.14   
1             1.806829e+18   jewelry.pendant  1.000000e+00   54.66   
2             1.806829e+18   jewelry.pendant  0.000000e+00   88.90   
3             1.806829e+18  jewelry.necklace  0.000000e+00  417.67   
4             1.806829e+18   jewelry.earring  1.000000e+00  102.27   
...                    ...               ...           ...     ...   
95905                  NaN              0.00  1.515916e+18     NaN   
95906         1.806829e+18               NaN  0.000000e+00  229.99   
95907         1.806829e+18   jewelry.earring  1.000000e+00  630.00   
95908         1.806829e+18   jewelry.earring  0.000000e+00   83.42   
95909         1.806829e+18   jewelry.pendant  1.000000e+00   97.12   

       1515915625207851155 Unnamed: 9     red  gold   diamond  
0             1.515916e+18        NaN  yellow  gold       NaN  
1             1.515916e+18          f   white  gold  sapphire  
2             1.515916e+18          f     red  gold   diamond  
3             1.515916e+18        NaN     red  gold  amethyst  
4             1.515916e+18        NaN     red  gold       NaN  
...                    ...        ...     ...   ...       ...  
95905                  NaN        NaN     NaN   NaN       NaN  
95906         1.515916e+18        NaN     red  gold       NaN  
95907         1.515916e+18          f   white  gold   diamond  
95908         1.515916e+18        NaN   white  gold    fianit  
95909         1.515916e+18        NaN     red  gold       NaN  

[95910 rows x 13 columns]

In [6]:
# Take a look at the first 5 rows
df.head()

2018-12-01 11:40:29 UTC  1924719191579951782  1842195256808833386  1  \
0  2018-12-01 17:38:31 UTC  1924899396621697920  1806829193678291446  1   
1  2018-12-02 13:53:42 UTC  1925511016616034733  1842214461889315556  1   
2  2018-12-02 17:44:02 UTC  1925626951238681511  1835566849434059453  1   
3  2018-12-02 21:30:19 UTC  1925740842841014667  1873936840742928865  1   
4  2018-12-02 22:09:34 UTC  1925760595336888995  1835566854827934449  1   

   1806829201890738522   jewelry.earring    0  561.51  1515915625207851155  \
0         1.806829e+18               NaN  NaN  212.14         1.515916e+18   
1         1.806829e+18   jewelry.pendant  1.0   54.66         1.515916e+18   
2         1.806829e+18   jewelry.pendant  0.0   88.90         1.515916e+18   
3         1.806829e+18  jewelry.necklace  0.0  417.67         1.515916e+18   
4         1.806829e+18   jewelry.earring  1.0  102.27         1.515916e+18   

  Unnamed: 9     red  gold   diamond  
0        NaN  yellow  gold       NaN  
1          f   white  gold  sapphire  
2          f     red  gold   diamond  
3        NaN     red  gold  amethyst  
4        NaN     red  gold       NaN

The dataset does not include column names and this will be fixed by using the column names provided in the data dictionary

In [7]:
# Include column names
df.columns = [
    "Order_Datetime",
    "Order ID",
    "Product_ID",
    "SKU_Quantity",
    "Category_ID",
    "Category",
    "Brand_ID",
    "Price_USD",
    "User_ID",
    "Gender",
    "Main_Color",
    "Main_Metal",
    "Main_Gem"]

In [8]:
# check if column names has been added
df.head(2)

Order_Datetime             Order ID           Product_ID  \
0  2018-12-01 17:38:31 UTC  1924899396621697920  1806829193678291446   
1  2018-12-02 13:53:42 UTC  1925511016616034733  1842214461889315556   

   SKU_Quantity   Category_ID         Category  Brand_ID  Price_USD  \
0             1  1.806829e+18              NaN       NaN     212.14   
1             1  1.806829e+18  jewelry.pendant       1.0      54.66   

        User_ID Gender Main_Color Main_Metal  Main_Gem  
0  1.515916e+18    NaN     yellow       gold       NaN  
1  1.515916e+18      f      white       gold  sapphire

In [9]:
# look at the number of rows and columns
df.shape

(95910, 13)

In [10]:
# calculate the number of null values
df.isnull().sum()

Order_Datetime        0
Order ID              0
Product_ID            0
SKU_Quantity          0
Category_ID        5352
Category           9933
Brand_ID           4785
Price_USD          5352
User_ID            5352
Gender            48167
Main_Color         7660
Main_Metal         5462
Main_Gem          34058
dtype: int64

The dataset contains lots of missing values will be treated subsequently

In [11]:
# check the data info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95910 entries, 0 to 95909
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Order_Datetime  95910 non-null  object 
 1   Order ID        95910 non-null  int64  
 2   Product_ID      95910 non-null  int64  
 3   SKU_Quantity    95910 non-null  int64  
 4   Category_ID     90558 non-null  float64
 5   Category        85977 non-null  object 
 6   Brand_ID        91125 non-null  float64
 7   Price_USD       90558 non-null  float64
 8   User_ID         90558 non-null  float64
 9   Gender          47743 non-null  object 
 10  Main_Color      88250 non-null  object 
 11  Main_Metal      90448 non-null  object 
 12  Main_Gem        61852 non-null  object 
dtypes: float64(4), int64(3), object(6)
memory usage: 9.5+ MB


In [12]:
# check for duplicates
df.duplicated().sum()

2589

The dataset contains 95,910 rows and 13 columns. It includes a mix of data types like objects, integers and floats. After taking a quick look, you can see that some columns have missing values and there are over 2,000 duplicate entries which are issues that have to be fixed going forward.

###### Data Cleaning

In [13]:
# convert the datetime column to date datatype
df['Order_Datetime'] =pd.to_datetime(df['Order_Datetime'], errors='coerce')

In [14]:
# confirm if conversion worked
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95910 entries, 0 to 95909
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype              
---  ------          --------------  -----              
 0   Order_Datetime  95910 non-null  datetime64[ns, UTC]
 1   Order ID        95910 non-null  int64              
 2   Product_ID      95910 non-null  int64              
 3   SKU_Quantity    95910 non-null  int64              
 4   Category_ID     90558 non-null  float64            
 5   Category        85977 non-null  object             
 6   Brand_ID        91125 non-null  float64            
 7   Price_USD       90558 non-null  float64            
 8   User_ID         90558 non-null  float64            
 9   Gender          47743 non-null  object             
 10  Main_Color      88250 non-null  object             
 11  Main_Metal      90448 non-null  object             
 12  Main_Gem        61852 non-null  object             
dtypes: datetime64[ns, UTC](1), floa

In [15]:
# drop duplicates
df.drop_duplicates(inplace=True)

In [16]:
# Fill categorical missing values with 'Unknown'
df['Category'].fillna('Unknown', inplace=True)

df['Brand_ID'].fillna('Unknown', inplace=True)

df['Main_Metal'].fillna('Unknown', inplace=True)

df['Main_Color'].fillna('unknown-color', inplace=True) # merge the unknown values with the unknown-color entry

df['Gender'].fillna('Unknown', inplace=True)

df['Main_Gem'].fillna('Unknown', inplace=True)

In [17]:
# Drop missing values in target Price
df = df.dropna(subset=['Price_USD'])

In [18]:
# Convert Brand_ID float entries to single values
df['Brand_ID'] = df['Brand_ID'].apply(lambda x: str(int(x)) if isinstance(x, (float, int)) and not pd.isna(x) else str(x))

In [19]:
# get counts of unique observations for each category variable
distinct_values=  df.select_dtypes(['object']).columns
for i in range(len(distinct_values)):
    print(df[distinct_values[i]].value_counts())

jewelry.earring       29047
jewelry.ring          26025
jewelry.pendant       13083
Unknown                9837
jewelry.bracelet       6129
jewelry.necklace       2335
jewelry.brooch         1286
jewelry.souvenir        170
electronics.clocks      165
jewelry.stud             12
Name: Category, dtype: int64
0          42919
1          31471
2           8838
Unknown     4675
4             65
3             58
6             56
5              7
Name: Brand_ID, dtype: int64
f          46687
Unknown    41043
m            359
Name: Gender, dtype: int64
red              67909
white            14755
yellow            3322
unknown-color     2101
black                2
Name: Main_Color, dtype: int64
gold        86618
silver       1362
Unknown       107
platinum        2
Name: Main_Metal, dtype: int64
diamond                28950
Unknown                27203
fianit                 11935
topaz                   6260
pearl                   3593
garnet                  2680
sapphire                1

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 88089 entries, 0 to 95909
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype              
---  ------          --------------  -----              
 0   Order_Datetime  88089 non-null  datetime64[ns, UTC]
 1   Order ID        88089 non-null  int64              
 2   Product_ID      88089 non-null  int64              
 3   SKU_Quantity    88089 non-null  int64              
 4   Category_ID     88089 non-null  float64            
 5   Category        88089 non-null  object             
 6   Brand_ID        88089 non-null  object             
 7   Price_USD       88089 non-null  float64            
 8   User_ID         88089 non-null  float64            
 9   Gender          88089 non-null  object             
 10  Main_Color      88089 non-null  object             
 11  Main_Metal      88089 non-null  object             
 12  Main_Gem        88089 non-null  object             
dtypes: datetime64[ns, UTC](1), floa

In [21]:
# statistical analysis of the numerical variables
df.describe()

Order ID    Product_ID  SKU_Quantity   Category_ID     Price_USD  \
count  8.808900e+04  8.808900e+04       88089.0  8.808900e+04  88089.000000   
mean   2.477630e+18  1.832195e+18           1.0  1.805950e+18    363.583797   
std    1.931240e+17  2.071288e+17           0.0  2.080101e+16    448.344088   
min    1.924899e+18  1.313551e+18           1.0  1.313678e+18      0.990000   
25%    2.373431e+18  1.806829e+18           1.0  1.806829e+18    143.700000   
50%    2.516754e+18  1.956664e+18           1.0  1.806829e+18    259.970000   
75%    2.637983e+18  1.956664e+18           1.0  1.806829e+18    431.370000   
max    2.719022e+18  2.541962e+18           1.0  1.806829e+18  34448.600000   

            User_ID  
count  8.808900e+04  
mean   1.512610e+18  
std    2.384248e+16  
min    1.313554e+18  
25%    1.515916e+18  
50%    1.515916e+18  
75%    1.515916e+18  
max    1.554297e+18

In [22]:
# statistical analysis of the categorical variables
df.describe(include="object")

Category Brand_ID Gender Main_Color Main_Metal Main_Gem
count             88089    88089  88089      88089      88089    88089
unique               10        8      3          5          4       31
top     jewelry.earring        0      f        red       gold  diamond
freq              29047    42919  46687      67909      86618    28950

- From the statistical summary, Price_USD is the only numerical column that makes sense. It has an minimum value of 0.99 and maximum of 30000. It is positively skewed as the average value of 353 is higher than the median of 253.
- Category has jewelry.earring as the most bought product.
- Target_Gender is mostly made up of Females
- Red seems to be the most frequent main_color.
- Gold is the most frequent main_metal.
- Diamond occurs most for main_gem.
- 0 seems to be the most frequent Brand_ID.
- Features like Order_ID, Product_ID, Category_ID and User_ID don't seem to be useful in this dataset and will likely be removed.